SaltXMLExporter throws PepperModuleException

_From @sdruskat on January 16, 2018 9:15_

- Ubuntu 16.04
- Pepper 3.1.1-SNAPSHOT (https://github.com/korpling/pepper/commit/0e20e6bbead899607022e9b3d497ebae91750504)

Simple Pepper workflow definition:

```
<?xml version='1.0' encoding='UTF-8'?>
<pepper-job id="some-title" version="1.0">
	<importer name="MyImporter" path="../mycorpus/"/>
	<exporter name="SaltXMLExporter" path="../sxml/">
	</exporter>
</pepper-job>
```
Result:
```
CONVERSION ENDED WITH ERRORS, REQUIRED TIME: 00:00:06.151 s
   Failed to process document by module 'ex1:SaltXMLExporter(SaltXMLExporter, 0.0)'. Nested exception was:  (PepperModuleException)
full stack trace:
org.corpus_tools.pepper.modules.exceptions.PepperModuleException: Failed to process document by module 'ex1:SaltXMLExporter(SaltXMLExporter, 0.0)'. Nested exception was: 
	at org.corpus_tools.pepper.core.PepperJobImpl.convert(PepperJobImpl.java:998)
	at org.corpus_tools.pepper.cli.PepperStarter.convert(PepperStarter.java:533)
	at org.corpus_tools.pepper.cli.PepperStarter.main(PepperStarter.java:1436)
Caused by: org.corpus_tools.salt.exceptions.SaltResourceException: Cannot store document graph to file '/home/stephan/sxml/corpus/document/doc_2.salt', because the opened stream is not closable. 
	at org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer.writeDocumentGraph(SaltXML10Writer.java:354)
	at org.corpus_tools.salt.util.SaltUtil.saveDocumentGraph(SaltUtil.java:610)
	at org.corpus_tools.salt.common.impl.SDocumentImpl.saveDocumentGraph(SDocumentImpl.java:189)
	at org.corpus_tools.salt.util.SaltUtil.saveCorpusGraph(SaltUtil.java:661)
	at org.corpus_tools.salt.util.SaltUtil.saveSaltProject(SaltUtil.java:781)
	at org.corpus_tools.pepper.modules.coreModules.SaltXMLExporter.start(SaltXMLExporter.java:140)
	at org.corpus_tools.pepper.core.ModuleControllerImpl$2.run(ModuleControllerImpl.java:289)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.ctc.wstx.exc.WstxIOException: Stream Closed
	at com.ctc.wstx.sw.BaseStreamWriter.flush(BaseStreamWriter.java:262)
	at org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer.writeDocumentGraph(SaltXML10Writer.java:351)
	... 11 more
Caused by: java.io.IOException: Stream Closed
	at java.io.FileOutputStream.writeBytes(Native Method)
	at java.io.FileOutputStream.write(FileOutputStream.java:326)
	at com.ctc.wstx.io.UTF8Writer.flush(UTF8Writer.java:100)
	at com.ctc.wstx.sw.BufferingXmlWriter.flush(BufferingXmlWriter.java:242)
	at com.ctc.wstx.sw.BaseStreamWriter.flush(BaseStreamWriter.java:260)
	... 12 more
```

_Copied from original issue: korpling/pepper#113_

Additional comment for @thomaskrause:

Further to what we have discussed, simply not closing the `XMLStreamWriter` doesn't work. It seems that it must be explicitly closed after all, otherwise we get a `ConcurrentModificationException`:

```
CONVERSION ENDED WITH ERRORS, REQUIRED TIME: 00:00:07.485 s
   Failed to process document by module 'ex1:SaltXMLExporter(SaltXMLExporter, 0.0)'. Nested exception was:  (PepperModuleException)
full stack trace:
org.corpus_tools.pepper.modules.exceptions.PepperModuleException: Failed to process document by module 'ex1:SaltXMLExporter(SaltXMLExporter, 0.0)'. Nested exception was: 
	at org.corpus_tools.pepper.core.PepperJobImpl.convert(PepperJobImpl.java:996)
	at org.corpus_tools.pepper.cli.PepperStarter.convert(PepperStarter.java:533)
	at org.corpus_tools.pepper.cli.PepperStarter.executeSingleCommand(PepperStarter.java:1195)
	at org.corpus_tools.pepper.cli.PepperStarter.runInteractive(PepperStarter.java:1157)
	at org.corpus_tools.pepper.cli.PepperStarter.main(PepperStarter.java:1390)
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:907)
	at java.util.ArrayList$Itr.next(ArrayList.java:857)
	at java.util.Collections$UnmodifiableCollection$1.next(Collections.java:1042)
	at org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer.writeDocumentGraph(SaltXML10Writer.java:409)
	at org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer.writeDocumentGraph(SaltXML10Writer.java:343)
	at org.corpus_tools.salt.util.SaltUtil.saveDocumentGraph(SaltUtil.java:610)
	at org.corpus_tools.salt.common.impl.SDocumentImpl.saveDocumentGraph(SDocumentImpl.java:189)
	at org.corpus_tools.salt.util.SaltUtil.saveCorpusGraph(SaltUtil.java:661)
	at org.corpus_tools.salt.util.SaltUtil.saveSaltProject(SaltUtil.java:781)
	at org.corpus_tools.pepper.modules.coreModules.SaltXMLExporter.start(SaltXMLExporter.java:140)
	at org.corpus_tools.pepper.core.ModuleControllerImpl$2.run(ModuleControllerImpl.java:289)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/salt-api/src/findbugs/excludeFilter.xml
+++ b/salt-api/src/findbugs/excludeFilter.xml
@@ -15,4 +15,8 @@
     <Class name="org.corpus_tools.salt.extensions.notification.SaltNotificationFactory"/>
     <Bug pattern="ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD"/>
   </Match>
+  <Match>
+    <Class name="org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer"/>
+    <Bug pattern="OBL_UNSATISFIED_OBLIGATION"/>
+  </Match>
 </FindBugsFilter>

--- a/salt-api/src/main/java/org/corpus_tools/salt/util/internal/persistence/SaltXML10Writer.java
+++ b/salt-api/src/main/java/org/corpus_tools/salt/util/internal/persistence/SaltXML10Writer.java
@@ -332,7 +332,9 @@ public class SaltXML10Writer implements SaltXML10Dictionary {
 	 */
 	public void writeDocumentGraph(SDocumentGraph graph) {
 		XMLStreamWriter xml = null;
-		try (OutputStream output = new FileOutputStream(path)) {
+		OutputStream output = null;
+		try {
+			output = new FileOutputStream(path);
 			xml = xmlFactory.createXMLStreamWriter(output, "UTF-8");
 
 			xml.writeStartDocument("1.0");
@@ -343,16 +345,29 @@ public class SaltXML10Writer implements SaltXML10Dictionary {
 			writeDocumentGraph(xml, graph);
 
 			xml.writeEndDocument();
-		} catch (XMLStreamException | IOException e) {
+		}
+		catch (XMLStreamException | IOException e) {
 			throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr() + "'. ", e);
-		} finally {
+		}
+		finally {
 			if (xml != null) {
 				try {
 					xml.flush();
 					xml.close();
-				} catch (XMLStreamException e) {
+				}
+				catch (XMLStreamException e) {
 					throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
-							+ "', because the opened stream is not closable. ", e);
+							+ "', because the opened XML stream is not closable. ", e);
+				}
+				if (output != null) {
+					try {
+						output.flush();
+						output.close();
+					}
+					catch (IOException e) {
+						throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
+						+ "', because the opened OutputStream is not closable. ", e);
+					}
 				}
 			}
 		}

--- a/salt-api/src/main/java/org/corpus_tools/salt/util/internal/persistence/SaltXML10Writer.java
+++ b/salt-api/src/main/java/org/corpus_tools/salt/util/internal/persistence/SaltXML10Writer.java
@@ -359,15 +359,15 @@ public class SaltXML10Writer implements SaltXML10Dictionary {
 					throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
 							+ "', because the opened XML stream is not closable. ", e);
 				}
-				if (output != null) {
-					try {
-						output.flush();
-						output.close();
-					}
-					catch (IOException e) {
-						throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
-						+ "', because the opened OutputStream is not closable. ", e);
-					}
+			}
+			if (output != null) {
+				try {
+					output.flush();
+					output.close();
+				}
+				catch (IOException e) {
+					throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
+					+ "', because the opened OutputStream is not closable. ", e);
 				}
 			}
 		}

--- a/salt-api/src/main/java/org/corpus_tools/salt/util/internal/persistence/SaltXML10Writer.java
+++ b/salt-api/src/main/java/org/corpus_tools/salt/util/internal/persistence/SaltXML10Writer.java
@@ -332,9 +332,7 @@ public class SaltXML10Writer implements SaltXML10Dictionary {
 	 */
 	public void writeDocumentGraph(SDocumentGraph graph) {
 		XMLStreamWriter xml = null;
-		OutputStream output = null;
-		try {
-			output = new FileOutputStream(path);
+		try (OutputStream output = new FileOutputStream(path);){
 			xml = xmlFactory.createXMLStreamWriter(output, "UTF-8");
 
 			xml.writeStartDocument("1.0");
@@ -345,31 +343,17 @@ public class SaltXML10Writer implements SaltXML10Dictionary {
 			writeDocumentGraph(xml, graph);
 
 			xml.writeEndDocument();
+			try {
+				xml.flush();
+				xml.close();
+			}
+			catch (XMLStreamException e) {
+				throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
+				+ "', because the opened XML stream is not closable. ", e);
+			}
 		}
 		catch (XMLStreamException | IOException e) {
 			throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr() + "'. ", e);
-		}
-		finally {
-			if (xml != null) {
-				try {
-					xml.flush();
-					xml.close();
-				}
-				catch (XMLStreamException e) {
-					throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
-							+ "', because the opened XML stream is not closable. ", e);
-				}
-			}
-			if (output != null) {
-				try {
-					output.flush();
-					output.close();
-				}
-				catch (IOException e) {
-					throw new SaltResourceException("Cannot store document graph to file '" + getLocationStr()
-					+ "', because the opened OutputStream is not closable. ", e);
-				}
-			}
 		}
 	}
 

--- a/salt-api/src/test/java/org/corpus_tools/salt/util/persistence/tests/Persist_SaltXML10_Test.java
+++ b/salt-api/src/test/java/org/corpus_tools/salt/util/persistence/tests/Persist_SaltXML10_Test.java
@@ -17,12 +17,17 @@
  */
 package org.corpus_tools.salt.util.persistence.tests;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
 
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
 import java.util.List;
 import java.util.Set;
 
@@ -36,6 +41,7 @@ import org.corpus_tools.salt.common.SCorpusGraph;
 import org.corpus_tools.salt.common.SDocument;
 import org.corpus_tools.salt.common.SDocumentGraph;
 import org.corpus_tools.salt.common.SaltProject;
+import org.corpus_tools.salt.core.SAnnotation;
 import org.corpus_tools.salt.samples.SampleGenerator;
 import org.corpus_tools.salt.tests.SaltTestsUtil;
 import org.corpus_tools.salt.util.Difference;
@@ -299,6 +305,55 @@ public class Persist_SaltXML10_Test {
 				loaded.getCorpusGraphs().get(0).getDocuments().get(2).getDocumentGraphLocation().toFileString());
 		assertEquals(outFolder + "/rootCorpus/subCorpus2/doc4.salt",
 				loaded.getCorpusGraphs().get(0).getDocuments().get(3).getDocumentGraphLocation().toFileString());
+	}
+	
+	/**
+	 * Tests persisting a SaltProject structure with {@link SAnnotation}s
+	 * on {@link SDocument}s.
+	 * 
+	 * The expected behaviour is that annotations on documents are persisted
+	 * in the corpus graph file, i.e. SaltProject.salt.
+	 */
+	@Test
+	public void testLoadStore_SaltProjectWithDocumentAnnotations() {
+
+		SaltProject project = SampleGenerator.createSaltProject();
+		for (SCorpusGraph cg : project.getCorpusGraphs()) {
+			for (SDocument d : cg.getDocuments()) {
+				d.createAnnotation("TEST", "DOCUMENT-ANNOTATION", d.getName());
+			}
+		}
+		for (SCorpusGraph cg : project.getCorpusGraphs()) {
+			for (SDocument d : cg.getDocuments()) {
+				assertNotNull(d.getAnnotation("TEST", "DOCUMENT-ANNOTATION"));
+				assertEquals(d.getAnnotation("TEST", "DOCUMENT-ANNOTATION").getValue_STEXT(), d.getName());
+			}
+		}
+
+		String outFolder = SaltTestsUtil.getTempTestFolder("/testLoadStore_SaltProjectWithSocumentAnnotations") + "/saltProject";
+		File tmpFile = new File(outFolder + "/" + SaltUtil.FILE_SALT_PROJECT);
+		SaltUtil.saveSaltProject(project, URI.createFileURI(tmpFile.getAbsolutePath()));
+		SaltProject loaded = SaltUtil.loadSaltProject(URI.createFileURI(tmpFile.getAbsolutePath()));
+
+		for (SCorpusGraph cg : loaded.getCorpusGraphs()) {
+			for (SDocument d : cg.getDocuments()) {
+				assertNotNull(d.getAnnotation("TEST", "DOCUMENT-ANNOTATION"));
+				assertEquals(d.getAnnotation("TEST", "DOCUMENT-ANNOTATION").getValue_STEXT(), d.getName());
+			}
+		}
+		
+		// Also check XML
+		List<String> lines = null;
+		try {
+			lines = Files.readAllLines(Paths.get(tmpFile.getAbsolutePath()));
+		}
+		catch (IOException e) {
+			fail("IOException");
+		}
+		assertTrue(lines.contains("		<labels xsi:type=\"saltCore:SAnnotation\" namespace=\"TEST\" name=\"DOCUMENT-ANNOTATION\" value=\"T::doc1\"/>"));
+		assertTrue(lines.contains("		<labels xsi:type=\"saltCore:SAnnotation\" namespace=\"TEST\" name=\"DOCUMENT-ANNOTATION\" value=\"T::doc2\"/>"));
+		assertTrue(lines.contains("		<labels xsi:type=\"saltCore:SAnnotation\" namespace=\"TEST\" name=\"DOCUMENT-ANNOTATION\" value=\"T::doc3\"/>"));
+		assertTrue(lines.contains("		<labels xsi:type=\"saltCore:SAnnotation\" namespace=\"TEST\" name=\"DOCUMENT-ANNOTATION\" value=\"T::doc4\"/>"));
 	}
 
 	/**

